### PR TITLE
Fix HTTP code for file request (static, assets, file)

### PR DESCRIPTION
**Pull request by micky2be**
# Description

- requesting for a non existing file, mostly through `file=path/filename.ext` would trigger a 500 error
- this update allows for a more smooth HTTP response like 404 or even 403
- simplified the `safe_join` for the use within gradio

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added a short summary of my change to the CHANGELOG.md
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by abidlabs** [comment]
Thanks @micky2be for making this fix! Generally looks good -- however, we should leave `gr.utils.abspath()` as is. The reason for that is because the `pathlib.Path.absolute()` path hasn't been well-tested in Python 3.7 and so we have avoided using it. You can [read (a lot) more about it here if you are interested](https://discuss.python.org/t/pathlib-absolute-vs-resolve/2573/13).

If you can revert that change, happy to review th rest.

**Comment by micky2be** [comment]
@abidlabs I see. Reverting as recommended.

File: gradio/routes.py
Comment (micky2be) [inline gradio/routes.py]: Not sure why you rolled back that change. It will break other part of the code which now rely on this.

**Comment by abidlabs** [comment]
Made some more tweaks since the `is_relative_to` method was introduced in Python 3.9, but `gradio` currently supports 3.7

**Comment by micky2be** [comment]
You didn't have to roll back. Your tweak is breaking the code, and will not trigger a proper 404.
I'm not using `Path.is_relative_to` in the code

**Comment by abidlabs** [comment]
You were using it here: 

>     if filename.is_absolute() or not is_relative_to(full_path.parent, directory):
>         raise HTTPException(403)

**Comment by micky2be** [comment]
`is_absolute` exist in 3.7

File: gradio/routes.py
Comment (micky2be) [inline gradio/routes.py:591]: @abidlabs this is what you are looking for

Comment (abidlabs) [reply gradio/routes.py:591]: But `is_relative_to` relies on the pathlib's Path's `relative_to` method internally. So it will not work on Python 3.7, right?

Comment (micky2be) [reply gradio/routes.py:591]: Yes, `relative_to ` was introduced a while ago

Comment (abidlabs) [reply gradio/routes.py:591]: Ah my apologies for the confusion

**Comment by abidlabs** [comment]
I'm re-reviewing the code. Regarding rewriting the `safe_join` method -- it's a critical method as it governs file access throughout all of gradio, so unless there is a strong reason to do it, I'm still not in favor of it. I especially do not want to use `pathlib.Path.absolute()` because, as I mentioned earlier, it is severely undertested in Python 3.7.

Would you be willing to leave the logic of `safe_join` as is and make the other changes outside of the method to raise the 403/404s as I had started to do?

**Comment by micky2be** [comment]
I'm not using `pathlib.Path.absolute()` at all.
Previous logic was a copy/paste of a library that does slightly different things (allow multiple arguments as path)
Here the code is much simpler and make sure `path` is not accessing files outside of `directory`.
That's why `test_route.py::TestRoutes::test_static_files_served_safely` is still passing

`safe_join` is used by the `/static/{path:path}` and `/assets/{path:path}` routes

**Comment by micky2be** [comment]
But if changing `safe_join` is such a deal breaker for you (_despite not doing much beside looking for absolute path and access outside of `directory`_), I'll update the code tomorrow

File: gradio/routes.py
Comment (micky2be) [inline gradio/routes.py:608]: This cover Lines 609 to 611 from previous code.
- check if `path` is an absolute path
- avoid `path` accessing files outside of `directory`

Basically forbidden access, hence the 403

File: gradio/routes.py
Comment (micky2be) [inline gradio/routes.py:602]: This will return an error instead of a 200 when accessing the root path of `directory`

File: gradio/routes.py
Comment (micky2be) [inline gradio/routes.py:611]: This check if we can actually return the requested file

**Review by micky2be — COMMENTED** [review]
Making sure the code is understood

File: gradio/routes.py
Comment (micky2be) [reply gradio/routes.py:602]: Something to add and doable is to check if `path` is requesting a folder and instead of a file

**Comment by abidlabs** [comment]
> But if changing safe_join is such a deal breaker for you (despite not doing much beside looking for absolute path and access outside of directory), I'll update the code tomorrow

Thanks for your patience @micky2be! I would strongly prefer that we do not change the logic of the `safe_join` method. There are small differences in how different OS-es handle path resolution and we've spent a lot of time in the past debugging these. I would rather not go through all of that testing again. 

If you can please refactor this so that the logic of the `safe_join` method is the same as before, I'd be happy to merge this in. If you would like to raise the HTTP errors inside the `safe_join()` method to reduce code duplication, that is fine. But we should use the same `os.path` module instead of `pathlib` and keep the overall logic the same.

**Comment by micky2be** [comment]
Alright! Hope that fits the need this time 😄
@abidlabs let me know if it looks good to you.
Not sure how to trigger the workflows.

**Comment by micky2be** [comment]
Interesting! It seems my new tests exposed an issue related to `pathlib.Path.resolve()` (fixed in Python 3.10) when trying to resolve the path of a non existing file on Windows.

Working on a fix for it right now.

**Comment by micky2be** [comment]
Fixed the issue with unresolved path (for non existing path) in `utils.abspath`
Also noticed `is_symlink` is not recursive so it would miss symlinks and would resolve it.
There was also a potential issue if an absolute path would be passed and be a symlink.

**Comment by abidlabs** [comment]
Cool thanks for making these fixes @micky2be! OOO but will review when I'm back in the office (cc @freddyaboulton or @aliabid94 if you guys wanna take a look in the meantime)

File: gradio/routes.py
Comment (freddyaboulton) [inline gradio/routes.py:319]: We should raise a 403 if abs_path is a directory

**Review by freddyaboulton — APPROVED** [review]
Thanks for the improvement @micky2be and for collaborating with @abidlabs ! 

Tested on deployed spaces and locally and it's working well. Left a small comment for improvement. Should be good to go after that

File: gradio/routes.py
Comment (micky2be) [reply gradio/routes.py:319]: That's really subjective. Could be 400, 403, or 404.
But we are not checking for directory here anyway.
Is it something you want me to add?

Comment (micky2be) [reply gradio/routes.py:319]: Good idea!

**Comment by freddyaboulton** [comment]
Thanks again @micky2be !

**Comment by micky2be** [comment]
Hey guys (@freddyaboulton @abidlabs)
As I was expending the battery of test of the routes for the sake of it (and for fun) I realized two things I'm not sure about:
- The upload path is absolute and create files outside of the application folder (default for `tempfile.mkdtem()`)
  - Windows: `C:\Users\<user>\AppData\Local\Temp\<hash>/`
  - Linux: `/tmp/<hash>/`
- Users have full access of any files of the application (in_app_dir)
  - Potential security risk for some

**Comment by abidlabs** [comment]
Hi @micky2be both of those are expected behavior, and the latter is mentioned here: https://gradio.app/sharing-your-app/#security-and-file-access

We may explicitly allow blacklisting files as  mentioned here: https://github.com/gradio-app/gradio/issues/3939

Thanks so much again for the contribution @micky2be and for reviewing @freddyaboulton!

**Pull request merged**
